### DISCUSSION TITLE
/microsoft-authentication-provider-visual-studio-code/

### DISCUSSION THREAD

**Discussion by comment-migrator** [estruyf/eliostruyf-comments#594] [Comments]
How to use the Microsoft auth provider in Visual Studio Code · Elio Struyf
        
[https://www.eliostruyf.com/microsoft-authentication-provider-visual-studio-code/](https://www.eliostruyf.com/microsoft-authentication-provider-visual-studio-code/)

**Comment by matthiasbigl** [comment]
Hello i had a lot of difficulty replicating this would you be able to provide the repo

**Reply by estruyf** [reply]
I don't have a repo for this; it is basically the above code I shared. However, you also need to have the appropriate Azure Entra App registration.

Where are you stuck?

**Reply by matthiasbigl** [reply]
i figured it out at the end thank you so much it was more of a problem with the webview communication. Used your libary and it worked flawless:) thank you I tipped you a coffee

**Reply by estruyf** [reply]
Happy to hear! Thanks for the coffee ☕️

**Comment by deepalijain** [comment]
Hi, thank you so much for this article! Is there a way to replicate this behavior for authenticating with Github? I've looked through this - https://github.com/microsoft/vscode/tree/main/extensions/github-authentication. 

I was wondering specifically if it's possible to get the refresh token instead of the regular token with Github as well.

**Reply by estruyf** [reply]
Hi @deepalijain, in case you want to use the GitHub Authentication Provider in your extension, you can use it as follows:

```ts
import { authentication } from 'vscode';
const githubAuth = await authentication.getSession('github', ['read:user'], { createIfNone: true });

console.log(`Access token`, githubAuth.accessToken);
```

If you use the GitHub Authentication Provider, you cannot retrieve the refresh token.

**Comment by burtonrodman** [comment]
Thanks so much for this information.  I've never had an easier experience authenticating an app.

For reference, I'm using AD B2C.  I had to check "Allow public client flows" and add this redirect uri (from an initial error message):
ms-appx-web://Microsoft.AAD.BrokerPlugin/<client-id>

**Comment by amitjoshi438** [comment]
In my extension I need to call different services with different scopes, this leads to multiple authentication prompts to user. I'm using {slient: true}, but I guess because of different scopes it still shows multiple auth prompts. Is there a better way to handle this?

**Reply by estruyf** [reply]
The only way would be to combine the scopes. That way, you only create one auth prompt instead of various prompts. As I'm unaware of your use case/requirements, it might be that this is not possible.
